samples/security/TokenLoginModule.java

```diff
@@ -0,0 +1,256 @@
+
+        if (debug) {
+            if (sharedState.size() > 0) {
+                debug("Commit: Shared state:");
+                sharedState.forEach((k, v) -> debug(String.format("Commit:   %s=%s", k, v)));
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean abort() throws LoginException {
+        // Clear state here, if any was created, particularly anything stored in
+        // the sharedState object or sensitive data stored in memory, like passwords.
+
+        debug("Abort: Clearing shared state");
+
+        sharedState.remove(USERNAME_KEY);
+
+        return true;
+    }
+
+    @Override
+    public boolean logout() throws LoginException {
+        // This method removes Principals, and removes/destroys credentials associated with the
+        // Subject during the commit operation. This method should not touch those Principals
+        // or credentials previously existing in the Subject, or those added by other LoginModules.
+
+        debug(String.format("Logout: Logging out %s", username));
+
+        if (!subject.isReadOnly()) {
+            // Example:
+            // subject.getPrincipals(MyPrincipal.class).clear();
+        }
+
+        return true;
+    }
+
+    // Retrieve the LoginResult from the Login Context.
+    private LoginResult getLoginResultFromCallback() throws LoginException, IOException, UnsupportedCallbackException {
+        LoginResultCallback loginResultCallback = new LoginResultCallback();
+        handler.handle(new Callback[] {
+            loginResultCallback
+        });
+        return loginResultCallback.getLoginResult();
+    }
+
+    // Retrieve the authentication token.
+    private AuthenticationToken getTokenFromCallback() throws LoginException {
+        AuthenticationTokenCallback authTokenCallback = new AuthenticationTokenCallback();
+        try {
+            handler.handle(new Callback[] {
+                authTokenCallback
+            });
+        }
+        catch (Exception e) {
+            throw new LoginException(e.getMessage());
+        }
+        return authTokenCallback.getAuthenticationToken();
+    }
+
+    // Validate the token and extract the username to add to shared state. If the token
+    // is found to be invalid, an exception is thrown.
+    //
+    // This method expects the token to be in JSON format:
+    //
+    // {
+    //   username: "joe",
+    //   nonce: "5171483440790326",
+    //   tokenExpires: "2017-04-20T15:48:49.187Z"
+    // }
+    //
+    private void processToken(String tokenData) throws JSONException, LoginException {
+
+        JSONObject json = new JSONObject(tokenData);
+
+        // Validate that the token hasn't expired.
+        ZonedDateTime expires = ZonedDateTime.parse(json.getString("tokenExpires"));
+        ZonedDateTime now = ZonedDateTime.now(expires.getZone());
+        if (expires.isBefore(now)) {
+            throw new LoginException("Token has expired");
```

Comment:
I was thinking that the reason for failure might help, but you're right. No need to do that.